File: app/src/main/java/org/wen/gank/base/MvpPresenter.java
package org.wen.gank.base;

import android.os.Bundle;

/**
 * created by Jiahui.wen 2017-07-14
 */
public class MvpPresenter<V extends MvpView> {

    private V view;

    protected void onCreate(Bundle savedInstanceState) {

    }

    protected void onSaveInstanceState(Bundle outState) {

    }

    protected void onDestroy() {

    }

    protected final void attachView(V view) {
        this.view = view;
        onAttachView(view);
    }

    protected void onAttachView(V view) {

    }

    protected final void detachView(boolean retainInstance) {
        this.view = null;
        onDetachView();
    }

    protected void onDetachView() {

    }

    public V getView() {
        return view;
    }

    public boolean isViewAttach() {
        return view != null;
    }
}

File: app/src/main/java/org/wen/gank/mvp/MvpPresenter.kt
package org.wen.gank.mvp

import android.os.Bundle

/**
 * created by Jiahui.wen 2017-07-14
 */
open class MvpPresenter<V : MvpView> {

  var view: V? = null
    private set

  val isViewAttach: Boolean
    get() = view != null

  @Suppress("UseExpressionBody", "UNUSED_PARAMETER")
  fun onCreate(savedInstanceState: Bundle?) {

  }

  @Suppress("UseExpressionBody", "UNUSED_PARAMETER")
  fun onSaveInstanceState(outState: Bundle) {

  }

  fun onDestroy() {

  }

  fun attachView(view: V) {
    this.view = view
    onAttachView(view)
  }

  @Suppress("MemberVisibilityCanBePrivate", "UNUSED_PARAMETER")
  protected fun onAttachView(view: V) {

  }

  @Suppress("MemberVisibilityCanBePrivate", "UNUSED_PARAMETER")
  fun detachView(retainInstance: Boolean) {
    this.view = null
    onDetachView()
  }

  @Suppress("MemberVisibilityCanBePrivate")
  protected fun onDetachView() {

  }
}
